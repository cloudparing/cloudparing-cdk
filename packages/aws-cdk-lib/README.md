# Cloudparing AWS CDK Construct Library

CDK constructs focused on monitoring and reducing cloud infrastructure costs.

## Installation

<details><summary><strong>TypeScript</strong></summary>

> https://www.npmjs.com/package/@cloudparing/aws-cdk-lib

In your `package.json`:

```json
{
  "dependencies": {
    "@cloudparing/aws-cdk-lib": "^0.0.0",

    // peer dependencies
    "aws-cdk-lib": "^2.65.0",
    "constructs": "^10.0.5"

    // ...your other dependencies...
  }
}
```
</details>

## Features

### Cost and Usage Report (CUR) v2 Report

```ts

    const curBucket =
      props.bucket ||
      new cdk.aws_s3.Bucket(this, `Cur2Bucket`, {
        bucketName: `cur2-${cdk.Aws.ACCOUNT_ID}-${cdk.Aws.REGION}`,
      });

    new Cur2ExportDefinition(this, `Cur2ExportDefinition`, {
      name: 'cur2-daily-csv',
      description: props.exportDescription,
      bucket: curBucket,
    });

```

### Legacy Cost and Usage Report (CUR)

An L2 construct and CDK stack for configuring the AWS Cost and Usage Report.

**NOTE: The CUR report must be deployed to us-east-1**

**Example using the ReportDefinition construct:**
```
    new ReportDefinition(this, `ReportDefinitionId`, {
      reportName: "CUR Report",
      s3Bucket: yourS3Bucket,
    });
```

**Example using the provided Stacks:**
```
import { CurStack, Cur2Stack } from '@cloudparing/aws-cdk-lib';
import * as cdk from 'aws-cdk-lib';

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new cdk.App();

new CurStack(app, 'curStack-dev', {
  env: devEnv,
  stackName: 'CurStack',
  reportName: 'cur-daily-csv',
});

new Cur2Stack(app, 'cur2Stack-dev', {
  env: devEnv,
  stackName: 'Cur2Stack',
  name: 'cur2-daily-csv',
});

app.synth();

```

Both the L2 construct and the stack provide sensible defaults which can be customized by setting props.

```
    new ReportDefinition(this, `ReportDefinitionId`, {
      reportName: "CUR Report",
      s3Bucket: yourS3Bucket,
      format: 'Parquet'
    });
```


## License

This project is licensed under the Apache-2.0 License.