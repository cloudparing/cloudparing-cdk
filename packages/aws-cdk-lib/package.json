{
  "name": "@cloudparing/aws-cdk-lib",
  "description": "The cloudparing CDK library",
  "repository": {
    "type": "git",
    "url": "github.com/cloudparing/cloudparing-cdk"
  },
  "scripts": {
    "build": "npx projen build",
    "bump": "npx projen bump",
    "compat": "npx projen compat",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "docgen": "npx projen docgen",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "package-all": "npx projen package-all",
    "package:js": "npx projen package:js",
    "package:python": "npx projen package:python",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "test": "npx projen test",
    "test:watch": "npx projen test:watch",
    "unbump": "npx projen unbump",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
    "projen": "npx projen"
  },
  "author": {
    "name": "James Ayvaz",
    "email": "james@cloudparing.net",
    "organization": false
  },
  "devDependencies": {
    "@types/aws-lambda": "^8.10.133",
    "@types/jest": "^29",
    "@types/node": "^18",
    "@typescript-eslint/eslint-plugin": "^6",
    "@typescript-eslint/parser": "^6",
    "aws-cdk-lib": "2.112.0",
    "constructs": "10.3.0",
    "eslint": "^8",
    "eslint-import-resolver-typescript": "^3.5.5",
    "eslint-plugin-import": "^2.27.5",
    "jest": "^29",
    "jest-junit": "^15",
    "jsii": "~5.2.0",
    "jsii-diff": "^1.84.0",
    "jsii-docgen": "^9.0.10",
    "jsii-pacmak": "^1.84.0",
    "jsii-rosetta": "~5.2.0",
    "projen": "^0.71.110",
    "ts-jest": "^29",
    "typescript": "^5.1.3"
  },
  "peerDependencies": {
    "aws-cdk-lib": "^2.112.0",
    "constructs": "^10.3.0"
  },
  "dependencies": {
    "@aws-sdk/client-bcm-data-exports": "^3.502.0",
    "@aws-sdk/client-s3": "^3.504.0",
    "aws-lambda": "^1.0.7"
  },
  "bundledDependencies": [
    "@aws-sdk/client-bcm-data-exports",
    "@aws-sdk/client-s3",
    "aws-lambda"
  ],
  "keywords": [
    "aws",
    "cdk",
    "cost",
    "cur"
  ],
  "main": "lib/index.js",
  "license": "Apache-2.0",
  "publishConfig": {
    "access": "public"
  },
  "version": "0.0.0",
  "jest": {
    "testMatch": [
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)"
    ],
    "clearMocks": true,
    "collectCoverage": true,
    "coverageReporters": [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text"
    ],
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": [
      "/node_modules/"
    ],
    "testPathIgnorePatterns": [
      "/node_modules/"
    ],
    "watchPathIgnorePatterns": [
      "/node_modules/"
    ],
    "reporters": [
      "default",
      [
        "jest-junit",
        {
          "outputDirectory": "test-reports"
        }
      ]
    ],
    "transform": {
      "^.+\\.[t]sx?$": [
        "ts-jest",
        {
          "tsconfig": "tsconfig.dev.json"
        }
      ]
    }
  },
  "types": "lib/index.d.ts",
  "stability": "stable",
  "jsii": {
    "outdir": "dist",
    "targets": {
      "python": {
        "distName": "cloudparing.aws-cdk-lib",
        "module": "cloudparing.aws_cdk_lib"
      }
    },
    "tsc": {
      "outDir": "lib",
      "rootDir": "src"
    }
  },
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"npx projen\"."
}
