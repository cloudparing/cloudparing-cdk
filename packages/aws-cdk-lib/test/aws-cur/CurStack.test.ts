import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CurStack } from '../../lib';

test('snapshot: CurStack', () => {
  const app = new App();
  const stack = new CurStack(app, 'test', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
    reportName: 'ReportDefinitionName',
  });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
