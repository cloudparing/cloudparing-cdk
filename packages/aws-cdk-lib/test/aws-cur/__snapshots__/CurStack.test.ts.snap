// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot: CurStack 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "testCurBucket43ED1B2B": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "cur-",
              Object {
                "Ref": "AWS::AccountId",
              },
              "-",
              Object {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testCurBucketPolicyE42EACB7": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "testCurBucket43ED1B2B",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetBucketAcl",
                "s3:GetBucketPolicy",
              ],
              "Condition": Object {
                "StringEquals": Object {
                  "aws:SourceAccount": Object {
                    "Ref": "AWS::AccountId",
                  },
                  "aws:SourceArn": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":cur:us-east-1:",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":definition/*",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "billingreports.amazonaws.com",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testCurBucket43ED1B2B",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "s3:PutObject",
              "Condition": Object {
                "StringEquals": Object {
                  "aws:SourceAccount": Object {
                    "Ref": "AWS::AccountId",
                  },
                  "aws:SourceArn": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":cur:us-east-1:",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":definition/*",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "billingreports.amazonaws.com",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "testCurBucket43ED1B2B",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "testReportDefinitiontestReportDefinitionResource65DECF67": Object {
      "DependsOn": Array [
        "testCurBucketPolicyE42EACB7",
      ],
      "Properties": Object {
        "Compression": "GZIP",
        "Format": "textORcsv",
        "RefreshClosedReports": true,
        "ReportName": "ReportDefinitionName",
        "ReportVersioning": "OVERWRITE_REPORT",
        "S3Bucket": Object {
          "Ref": "testCurBucket43ED1B2B",
        },
        "S3Prefix": "/",
        "S3Region": "us-east-1",
        "TimeUnit": "DAILY",
      },
      "Type": "AWS::CUR::ReportDefinition",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
